### PR TITLE
Emit lossless metrics rather than StatisticsSet from CloudWatchRecorder

### PR DESCRIPTION
**Pull request by kcnolan7**
This commit accomplishes two goals: first it refactors MetricRecorder's MetricContext in order to enforce the documented behavior that "[counts] are explicitly aggregated values, where only the total number of occurences in a context matter."

The second is to emit lossless metrics from the CloudWatchRecorder.

### CODE REVIEW

File: metrics-api/src/main/java/software/amazon/swage/metrics/MetricContext.java [hunk 1]
```diff
@@ -53,9 +53,10 @@ default void record(Metric label, Number value, Unit unit) {
      * is often used to record occurrences of an event, such as the number of times
      * a method is called, an error occurred, or the amount of data sent.
      * <p/>
-     * Changes to the count are not timestamped as only the total value of all
-     * counts for a metric have any meaning. If the individual change needs to
-     * be tracked, it should be recorded as a gauged event.
+     * Changes to the count are timestamped with the time of context closure,
```
Comment (jerpowers) [inline metrics-api/src/main/java/software/amazon/swage/metrics/MetricContext.java]: This adds an explicit expectation/dependency on context closure to the context - now it assumes that every context will always be closed.  Is this appropriate?

File: metrics-core/src/main/java/software/amazon/swage/metrics/record/cloudwatch/CloudWatchRecorder.java [hunk 2]
```diff
@@ -14,23 +14,15 @@
...
 import java.time.Instant;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.concurrent.Executors;
-import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.ThreadLocalRandom;
-import java.util.concurrent.TimeUnit;
+import java.util.*;
```
Comment (jerpowers) [inline metrics-core/src/main/java/software/amazon/swage/metrics/record/cloudwatch/CloudWatchRecorder.java]: Do not use wildcard imports

File: metrics-api/src/main/java/software/amazon/swage/metrics/record/MetricRecorder.java [hunk 3]
```diff
@@ -133,11 +137,12 @@ public void record(Metric label, Number value, Unit unit, Instant time) {
 
             @Override
             public void count(Metric label, long delta) {
-                MetricRecorder.this.count(label, delta, context);
+                counts.compute(label, (x, previous) -> (previous == null) ? delta : previous + delta);
```
Comment (jerpowers) [inline metrics-api/src/main/java/software/amazon/swage/metrics/record/MetricRecorder.java]: So calling count now no longer actually records a metric, but stores and aggregates (regardless of implementation).  Not sure this is good.

It should be up to the recorder implementation do decide if it wants to aggregate things before sending.

File: metrics-api/src/main/java/software/amazon/swage/metrics/record/MetricRecorder.java [hunk 4]
```diff
@@ -121,6 +123,8 @@ public TypedMap attributes() {
     public MetricContext context(TypedMap attributes) {
         R context = newRecorderContext(attributes);
         return new MetricContext() {
+            private Map<Metric,Long> counts = new ConcurrentHashMap<>();
```
Comment (jerpowers) [inline metrics-api/src/main/java/software/amazon/swage/metrics/record/MetricRecorder.java]: See below - I don't believe this belongs here.  Aside from the separation of concerns thing, there is now a map created and stored for every metric context, even if said context never uses it.

File: metrics-api/src/main/java/software/amazon/swage/metrics/record/MetricRecorder.java [hunk 5]
```diff
@@ -172,29 +177,28 @@ protected abstract void record(
             R context);
 
     /**
-     * Count the increase or decrease of a metric in the given context.
+     * Record the value of a count metric at the closure of a context.
```
Comment (jerpowers) [inline metrics-api/src/main/java/software/amazon/swage/metrics/record/MetricRecorder.java]: See above comments - I don't think tying this to close is the right approach.

File: metrics-core/src/main/java/software/amazon/swage/metrics/record/cloudwatch/CloudWatchRecorder.java [hunk 6]
```diff
@@ -79,7 +71,9 @@
 
     private static final Logger log = LogManager.getLogger(CloudWatchRecorder.class);
 
-    // Count of metric datum to batch together in one CloudWatch call
+    // Count of metric data to batch together in one CloudWatch call
```
Comment (jerpowers) [inline metrics-core/src/main/java/software/amazon/swage/metrics/record/cloudwatch/CloudWatchRecorder.java]: data vs datum

Which is right?  It is impossible to knows.

File: metrics-core/src/main/java/software/amazon/swage/metrics/record/cloudwatch/CloudWatchRecorder.java [hunk 7]
```diff
@@ -454,19 +452,21 @@ private void sendAggregatedData() {
 
         // Grab all the current aggregated attributes, resetting
         // the aggregator to empty in the process
-        List<MetricDatum> metricData = aggregator.flush();
-        if(metricData.isEmpty()) {
+        List<MetricDatum> metrics = new ArrayList<>();
```
Comment (jerpowers) [inline metrics-core/src/main/java/software/amazon/swage/metrics/record/cloudwatch/CloudWatchRecorder.java]: Small nitpick: leaving name as 'metricData' makes the diff a little less noisy.

**Comment by kcnolan7** [comment]
> Is sending every metric event to CloudWatch acceptable? The aggregation is there to avoid excessive data and the excessive monetary charges that go along with it - is non-aggregated always the right approach now?

I'd make two points here:
1) In the base metrics-core repository, the recorders should really be considered examples rather than fully-fledged 'prod-ready' projects.  Simplifying this recorder to record all datapoints isn't so much a decision that non-aggregated is "always" the right choice so much as a completely reasonable choice.

2) For services recording less than 3000 metrics per second (150 tps limit on putMetricData calls, 20 'datums' per request = 3000 metrics per second limit), publishing lossless metrics is generally preferable in order to leverage percentiles in dashboards and alarms.

>  If the recorder implementation wants to aggregate before sending, cool, but it shouldn't be forced. Putting the aggregation in there changes the contract of the count method from 'pass stuff along to recorder' to 'do stateful things'.

The documentation *currently* stipulates that counts are "explicitly aggregated values, where only the total number of occurrences in a context matter."  By not implementing this in the abstract MetricRecorder, every implementation is responsible for getting this right (which the CloudWatchRecorder does not).  Further, if a count is just a guaged integer metric, what differentiates counts vs records at all?

File: metrics-core/src/main/java/software/amazon/swage/metrics/record/cloudwatch/CloudWatchRecorder.java (hunk 6 above)
Comment (markuspeloquin) [reply metrics-core/src/main/java/software/amazon/swage/metrics/record/cloudwatch/CloudWatchRecorder.java]: Count of things to batch together...

'data' is right

**Comment by kcnolan7** [comment]
This PR conflates two different actions: refactoring Context#count (which I still advocate for) and emitting lossless metrics (which I also still advocate for, but less strongly).  It's probably best to consider these two discussions separately.

**Pull request closed without merging**